security-utils/src/main/java/com/yahoo/security/KeyId.java

```diff
@@ -0,0 +1,89 @@
+// Copyright Yahoo. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.security;
+
+import java.util.Arrays;
+import java.util.Objects;
+
+import static com.yahoo.security.ArrayUtils.fromUtf8Bytes;
+import static com.yahoo.security.ArrayUtils.toUtf8Bytes;
+
+/**
+ * Represents a named key ID comprising an arbitrary (but length-limited)
+ * sequence of valid UTF-8 bytes.
+ *
+ * @author vekterli
+ */
+public class KeyId {
+
+    // Max length MUST be possible to fit in an unsigned byte; see SealedSharedKey token encoding/decoding.
+    public static final int MAX_KEY_ID_UTF8_LENGTH = 255;
+
+    private final byte[] keyIdBytes;
+
+    private KeyId(byte[] keyIdBytes) {
+        if (keyIdBytes.length > MAX_KEY_ID_UTF8_LENGTH) {
+            throw new IllegalArgumentException("Key ID is too large to be encoded (max is %d, got %d)"
+                                               .formatted(MAX_KEY_ID_UTF8_LENGTH, keyIdBytes.length));
+        }
+        verifyByteStringRoundtripsAsValidUtf8(keyIdBytes);
+        this.keyIdBytes = keyIdBytes;
+    }
+
+    /**
+     * Construct a KeyId containing the given sequence of bytes.
+     *
+     * @param keyIdBytes array of valid UTF-8 bytes. May be zero-length, but not null.
+     *                   Note: to avoid accidental mutations, the key bytes are deep-copied.
+     * @return a new KeyId instance
+     */
+    public static KeyId ofBytes(byte[] keyIdBytes) {
+        Objects.requireNonNull(keyIdBytes);
+        return new KeyId(keyIdBytes.clone());
```

Comment:
Use `Arrays.copyOf()`. `Object.clone()` is a sad part of Java's history and should never be used...

Reply:
Array copying fixed to be more cheerful, less sad